rfcs/text/0428-extending-types.md

```diff
@@ -0,0 +1,925 @@
+
+## The `additionalProperties` problem
+
+In the proposed [Versioning RFC](./0408-versioning-types.md) for the type system, having `{ "additionalProperties": false }` for all schemas is an assumption made for determining type compatibility, which means that any supertype will not validate against a subtype that adds properties if it receives all properties of a subtype instance. For example, if we supply the `Employee` instance from above to a `Person`, it will receive properties that are considered `additionalProperties` (the `occupation` property is not present on `Person`).
+
+The assumption that we can select/project parts of a subtype that make up a supertype is essential for keeping strictness in JSON Schemas.
+
+We propose slight modifications to how `{ "additionalProperties": true }` and `{ "unevaluatedProperties": false }` behave and may be used within our type extension system to let supertypes keep strictness while allowing composition.
+
+Concrete examples of how JSON Schema breaks with these validation constraints are shown in the [Reference-level explanation](#problems-with-unevaluatedproperties).
+
+## Defining extended entity types
+
+Extended types will be defined with conventional JSON Schema syntax: the `allOf` keyword. An entity type can extend another entity type by adding a versioned URI reference to the root-level `allOf` array.
+A [versioned URI](https://github.com/blockprotocol/blockprotocol/blob/main/rfcs/text/0408-versioning-types.md#type-uris) is used so that subtypes aren't automatically updated (and potentially invalidated) when the supertype is updated.
+
+As extended types can extend other extended types we must also make sure that there are no cycles within the type hierarchy, as it makes types difficult to resolve/reason about and could lead to unpredictable behavior.
+
+## Defining entity type duplication
+
+When looking for public types, if extending an entity type is insufficient, an alternative is to duplicate the type so that individual properties can be removed or overridden. Duplicated types become new complete, standalone types.
+
+Type duplication, unlike extended types, does not imply that instances of the original type can be substituted by instances of the duplicated type because the types may be incompatible. A mapping may be used to persist some of of compatibility (which is yet to be defined within the type system).
+
+Because of the behavior of type duplication, the Block Protocol wouldn't have to change much to support duplicating entity types, as it can be considered literal copying. Although having more explicit metadata about type duplication does, however, enable provenance for duplicated types. As such, any duplicated type should specify what origin entity type they are duplicates of through a new `forkedFrom` keyword. This keyword is used in the duplicated entity type schema as a versioned URI reference to the type which was duplicated.
```

Comment:
I think we should actually remove this. I'm not sure of the value of storing the metadata inline inside the schema, 
especially in the long run. 

If the metadata actually did prove useful and people relied on it for certain logic flows then I think we'd need to worry about how easy it is to spoof, I could just throw "forkedFrom" with any URI I like even if the two things are 100% unrelated. To be clear, I don't think there's actually much risk from people doing so, but that's because I think this metadata holds very little value.

If some big type aggregator / source / manager wants to maintain separate metadata then I think that's perfectly sensible and viable, but I don't think it makes sense to include it within the schema. And it adds another point for a (weak) dependency on another type.

Reply:
Removed in d2903f556d80e41fb21885e6c882ce114ac5dca3